### PR TITLE
MVTLayer: add onViewportLoad callback

### PR DESCRIPTION
**Pull request by AdriSolid**
#### Background
There are cases where it is necessary to retrieve all the visible tiles in the viewport, this way it is possible to get all `Tile2DHeader` info and make (post)operations.

The new `getVisibleTiles` method proposed returns `selectedTiles` from `tileset`.

```javascript
// Accessing via onViewportChange
const onViewportChange = e => {
  const visibleTiles = e.getVisibleTiles();
  visibleTiles.forEach(tile => {
   // tile.data;
   // tile.dataInWorldCoordinates;
  });

const layer = new MVTLayer({
  id: "..."
  data: "..."
  onViewportChange: debounce(onViewportChange, 500)
})

// Accessing via layer
layer.getVisibleTiles();
```

#### Change List
- [x] `dataInWorldCoordinates` getter - `Tile2DHeader`
- [x]  `transformToWorld` member - `Tile2DHeader`
- [x] `getVisibleTiles` method - `MVTLayer`
- [x] Cache for tile data transformation
- [x] Unit tests

### CODE REVIEW

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 1]
```diff
@@ -130,25 +130,29 @@ tileJSON (Object) - the tileJSON fetched
 
 ##### `onViewportChange` (Function, optional)
 
-`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded. A function (`getRenderedFeatures`) to calculate the features rendered in the current viewport is passed as a JSON to this callback function.
+`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded.
 
+Functions passed as a JSON to this callback function:
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/mvt-layer.md:135]: This is duplicate of "receives arguments" below

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 2]
```diff
@@ -161,6 +165,26 @@ new MVTLayer({
 })
 ```
 
+* `getVisibleTiles` (async Function)
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/mvt-layer.md:168]: `getVisibleTiles` (Function)

Arguments: ...

Returns: a promise that resolves to ...

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 3]
```diff
@@ -216,12 +216,24 @@ export default class MVTLayer extends TileLayer {
     return renderedFeatures;
   }
 
+  async getVisibleTiles(tileCoords = 'local') {
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:219]: Why is `tileCoords` required? You can just deduce the coordinate system from information in this layer.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 4]
```diff
@@ -267,5 +279,27 @@ function transformTileCoordsToWGS84(object, tile, viewport) {
   return feature;
 }
 
+function conversionToWGS84isRequired(tileCoords, isWGS84) {
+  return tileCoords === 'wgs84' && !isWGS84;
+}
+
+function updateTilesWithWGS84Coords(tiles, tilesData, viewport) {
+  tiles.forEach((tile, idx) => {
+    tile.dataWithWGS84Coords = tilesData[idx].map(object =>
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:288]: - The tile instance should never be mutated, let alone conditionally. Bear in mind that MVTLayer is not the only user of TileLayer, while the WGS84 issue is MVT specific. This will also introduce a lot of issues when we move to TypeScript. You can add a method `get dataInWorldCoordinates()` to the `Tile2DHeader` class.
- This seems to be recalculated every time you call `getVisibleTiles`

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 5]
```diff
@@ -267,5 +279,27 @@ function transformTileCoordsToWGS84(object, tile, viewport) {
...
+}
+
+async function convertTilesToWGS84(tiles, viewport) {
+  const tilesData = await Promise.all(
+    tiles.map(async tile => {
+      return (await tile.data) || [];
+    })
+  );
+
+  updateTilesWithWGS84Coords(tiles, tilesData, viewport);
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:301]: You can access the loaded data synchronously via `tile.content` once it's loaded. No need to construct a separate list.

**Comment by Pessimistress** [comment]
To expand on the `Tile2DHeader` change:

- add a `get dataInWorldCoordinates()` to `Tile2DHeader`. By default it returns the same thing as `data`
- add a member `transformToWorld` to `Tile2DHeader`. If specified, when `dataInWorldCoordinates` is first called, return `this.transformToWorld(this, this.content)` and cache the result.
- MVTLayer should add its custom transform to all selected tiles after each state update.

**Comment by AdriSolid** [comment]
@Pessimistress Thanks for your comments!

Just updated the logic following your pseudo code. Not sure about one thing related with 'cache', when you say:
`If specified, when dataInWorldCoordinates is first called, return this.transformToWorld(this, this.content) and **cache** the result.`

Are you talking about in creating a cache (like a js `Map`) in `Tile2DHeader`?

File: modules/geo-layers/src/mvt-layer/mvt-layer.js (hunk 3 above)
Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js]: Let's say `resolution` is = `'local'`, but user needs the tile data in WGS84 :roll_eyes:

**Comment by ibgreen** [comment]
I am bit late to this discussion, and I am more expert on 3D tiles than 2D tiles so I am making some assumptions and my comments may be slightly off the mark...

This type of methods (e.g. `getVisibleTiles`) should definitely be provided on the `Tileset2D` class. It is basically the job of the Tileset2D class to do this calculation based on a viewport (at least this is true for `Tileset3D`).

If that is true for `Tileset2D` then we can ask ourselves if Tileset methods should also be present on the Layer. How about just making the tileset available?

Then the code would looke like
```
layer.getTileset().getVisibleTiles()
```

Or
```
layer.tileset.getVisibleTiles()
```

This way all methods on the tileset are available to anyone that has a handle to the layer.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js (hunk 4 above)
Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js]: Didn't know about the migration to TS :+1:

File: docs/upgrade-guide.md [hunk 6]
```diff
@@ -24,8 +24,9 @@ A legacy field `pickingInfo.lngLat` has been removed. Use `pickingInfo.coordinat
 
 ### MVTLayer
 
-- `onHover` and `onClick` callbacks now throw the `info.object` feature coordinates in WGS84 standard.
-
+- `MVTLayer`'s picked info now yields the `info.object` feature coordinates in WGS84 standard.
+- `getRenderedFeatures` method in `MVTLayer`: calculates the features rendered in the current viewport.
```
Comment (Pessimistress) [inline docs/upgrade-guide.md:28]: Upgrade Guide should only contain depredations and breaking changes.

File: modules/geo-layers/src/mvt-layer/coordinate-transform.js [hunk 7]
```diff
@@ -45,12 +45,5 @@ export function transform(geometry, bbox, viewport) {
...
 
-  return {
-    ...geometry,
-    coordinates: availableTransformations[geometry.type](
-      geometry.coordinates,
-      projectedBbox,
-      viewport
-    )
-  };
+  return availableTransformations[geometry.type](geometry.coordinates, projectedBbox, viewport);
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/coordinate-transform.js:48]: These changes do not belong in this PR. Please rebase master.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 8]
```diff
@@ -216,12 +217,37 @@ export default class MVTLayer extends TileLayer {
     return renderedFeatures;
   }
 
+  getVisibleTiles(tileCoords = 'local') {
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:220]: Why is `tileCoords` needed?

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 9]
```diff
@@ -216,12 +217,37 @@ export default class MVTLayer extends TileLayer {
...
+
+  _convertTilesToWGS84(tiles, viewport) {
+    const {tileIdsCache} = this.state;
+
+    tiles.forEach(tile => {
+      const tileId = `${tile.x},${tile.y},${tile.z}`;
+
+      if (!tileIdsCache.has(tileId)) {
+        tileIdsCache.add(tileId);
+        tile.transformToWorld = () => updateTileWithWGS84Coords(tile, tile.content, viewport);
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:239]: This should be added when the tile is created, not when `getVisibleTiles` is called. Tiles can be accessed in other callbacks, e.g. `onTileLoad`, `renderSubLayers`

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 10]
```diff
@@ -216,12 +217,37 @@ export default class MVTLayer extends TileLayer {
...
+  }
+
+  _convertTilesToWGS84(tiles, viewport) {
+    const {tileIdsCache} = this.state;
+
+    tiles.forEach(tile => {
+      const tileId = `${tile.x},${tile.y},${tile.z}`;
+
+      if (!tileIdsCache.has(tileId)) {
+        tileIdsCache.add(tileId);
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:238]: What is the purpose of this? Just check if `transformToWorld` is assigned?

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 11]
```diff
@@ -267,5 +293,17 @@ function transformTileCoordsToWGS84(object, tile, viewport) {
   return feature;
 }
 
+function conversionToWGS84isRequired(tileCoords, isWGS84) {
+  return tileCoords === 'wgs84' && !isWGS84;
+}
+
+function updateTileWithWGS84Coords(tile, content, viewport) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:300]: You are not updating the tile, just transforming the content. Name the function accordingly.

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 12]
```diff
@@ -17,12 +18,22 @@ export default class Tile2DHeader {
...
+    this.transformToWorld = null;
   }
 
   get data() {
     return this._isLoaded ? this.content : this._loader;
   }
 
+  get dataInWorldCoordinates() {
+    if (this.transformToWorld) {
+      return this.transformToWorld(this, this.content);
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:31]: - This is still recalculating all the transforms every time this function is called. Content and transformToWorld are not expected to change.
- Not handling the case when content is still loading.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js (hunk 8 above)
Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:220]: Let's say `viewport.resolution` is = `'local'`, but user needs the tile data in WGS84 right?

Comment (Pessimistress) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:220]: If you want the local coords, you access `tile.data`. If you want WGS84, you access `tile.dataInWorldCoordinates`. Both should always be available in any tile instance.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js (hunk 10 above)
Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:238]: Indeed, checking if calculation is done for a certain tile, avoiding unnecessary calculations. We can even 
simplify it:

```javascript
if (!tile.transformToWorld ) {
  tile.transformToWorld = () => updateTileWithWGS84Coords(tile, tile.content, viewport);
}
```

Comment (Pessimistress) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:238]: > avoiding unnecessary calculations

Assigning `transformToWorld` is not doing any actual calculation.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js (hunk 9 above)
Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:239]: Should we move the calculation out of `MVTLayer`?

Comment (AdriSolid) [reply modules/geo-layers/src/mvt-layer/mvt-layer.js:239]: Better to move it inside `getTileData` `MVTLayer` method?

**Comment by AdriSolid** [comment]
@Pessimistress still not sure if calling `convertTilesToWGS84` [here](https://github.com/visgl/deck.gl/blob/54eb96163c1b021a419dafda5021993a66889b46/modules/geo-layers/src/mvt-layer/mvt-layer.js#L238) is the right place, any advice?

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 13]
```diff
@@ -130,25 +130,26 @@ tileJSON (Object) - the tileJSON fetched
 
 ##### `onViewportChange` (Function, optional)
 
-`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded. A function (`getRenderedFeatures`) to calculate the features rendered in the current viewport is passed as a JSON to this callback function.
+`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded.
 
+Remarks
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/mvt-layer.md:135]: Move to the end of this section

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 14]
```diff
@@ -216,12 +216,29 @@ export default class MVTLayer extends TileLayer {
     return renderedFeatures;
   }
 
+  getVisibleTiles() {
+    return this.state.tileset.selectedTiles;
+  }
+
+  _convertTilesToWGS84(tiles, viewport) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:223]: This method is not performing any conversion. Call it something like "setupTileTransform"

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 15]
```diff
@@ -216,12 +216,29 @@ export default class MVTLayer extends TileLayer {
...
+        return [];
+      };
+    });
+  }
+
   _onViewportChange() {
     const {onViewportChange} = this.props;
     if (onViewportChange) {
       const {viewport} = this.context;
+      this._convertTilesToWGS84(this.state.tileset.selectedTiles, viewport);
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:238]: You are still adding the transform conditionally (only when the user specifies `onViewportChange`. Move it to after `super.updateState`

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 16]
```diff
@@ -17,12 +18,29 @@ export default class Tile2DHeader {
 
     this.onTileLoad = onTileLoad;
     this.onTileError = onTileError;
+
+    this.transformToWorld = null;
+    this._transformToWorldDataCache = [];
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:23]: Use `null`

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 17]
```diff
@@ -17,12 +18,29 @@ export default class Tile2DHeader {
...
   get data() {
     return this._isLoaded ? this.content : this._loader;
   }
 
+  get dataInWorldCoordinates() {
+    const data = this._isLoaded ? this.content : this._loader;
+
+    if (this.transformToWorld) {
+      if (!this._transformToWorldDataCache.length) {
+        this._transformToWorldDataCache = this.transformToWorld(data);
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:35]: If you pass a promise to `transformToWorld` then `transformToWorld` needs to be an async function.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 18]
```diff
@@ -216,12 +217,24 @@ export default class MVTLayer extends TileLayer {
     return renderedFeatures;
   }
 
+  getVisibleTiles() {
+    return this.state.tileset.selectedTiles;
+  }
+
+  _setupTileTransform(tiles, viewport) {
+    tiles.forEach(tile => {
+      tile.transformToWorld = data =>
```
Comment (Pessimistress) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:226]: Don't create a new function if `tile.transformToWorld` has already been set

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 19]
```diff
@@ -17,12 +18,27 @@ export default class Tile2DHeader {
...
   }
 
   get data() {
     return this._isLoaded ? this.content : this._loader;
   }
 
+  get dataInWorldCoordinates() {
+    if (this.transformToWorld) {
+      if (this._transformToWorldDataCache === null) {
+        this._transformToWorldDataCache = this.content ? this.transformToWorld(this.content) : null;
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:33]: You need to await data if it's still loading. Currently this getter returns null with transform assigned, and a promise if it's not.

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 20]
```diff
@@ -17,12 +18,29 @@ export default class Tile2DHeader {
...
 
+  get dataInWorldCoordinates() {
+    if (this.transformToWorld) {
+      if (this._transformToWorldDataCache === null) {
+        this._transformToWorldDataCache = this.content
+          ? this.transformToWorld(this.content)
+          : this.content;
+      }
+
+      return this._isLoaded ? this._transformToWorldDataCache : this._loader;
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:38]: This is still incorrect. Please add a test case.

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 21]
```diff
@@ -130,25 +130,22 @@ tileJSON (Object) - the tileJSON fetched
 
 ##### `onViewportChange` (Function, optional)
 
-`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded. A function (`getRenderedFeatures`) to calculate the features rendered in the current viewport is passed as a JSON to this callback function.
-
+`onViewportChange` called when the viewport changes or all tiles in the current viewport have been loaded.
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/mvt-layer.md:133]: Suggestion: since you are also exposing these functions as layer methods, the callback can just be called with `{layer: this}`, then the user can cal `layer.getVisibleTiles`.

Either way the new methods should be documented.

**Comment by AdriSolid** [comment]
@Pessimistress I think I don't see the point.. Is that flow logic for you?
```javascript
get dataInWorldCoordinates() {
    if (this.transformToWorld) {
      // Cache is not set and data is loaded
      if (this._transformToWorldDataCache === null && this._isLoaded) {
        // Transform, set cache and return data
        this._transformToWorldDataCache = this.transformToWorld(this.content);
        return this._transformToWorldDataCache;
      }
      // Cache is not set, return loader
      return this._transformToWorldDataCache || this._loader;
    }
    // default
    return this._isLoaded ? this.content : this._loader;
  }
```

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 22]
```diff
@@ -117,38 +117,40 @@ On top of the [default options](/docs/api-reference/core/layer.md#loadoptions),
...
 - [MVTLoader](https://loaders.gl/modules/mvt/docs/api-reference/mvt-loader)
 
+## Methods
 
-### Callbacks
+##### `getVisibleTiles`
+
+`getVisibleTiles` retrieves the visible tiles in the current viewport.
+
+## Callbacks
```
Comment (Pessimistress) [inline docs/api-reference/geo-layers/mvt-layer.md:126]: The callbacks are also layer properties. Move the methods out of the properties section.

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 23]
```diff
@@ -65,3 +65,42 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+  const onTileError = () => null;
+
+  const tile = new Tile2DHeader({onTileLoad, onTileError});
+  tile.isSelected = true;
+  tile.transformToWorld = () => TILE_DATA_VALUES.transformed;
+
+  const loader = tile._loadData(getTileData, requestScheduler);
+  await loader;
+
+  if (tile.isLoaded) {
```
Comment (Pessimistress) [inline test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:87]: You are not testing the case where the tile is still loading. Your `dataInWorldCoordinates` getter does not return a promise that resolves to the correct coordinates.

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 24]
```diff
@@ -65,3 +67,104 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+test('Tile2DHeader#transformToWorld', async t => {
+  const [tile, getTileData] = createTile2DHeaderWithMockedData(LOCAL_TILE_COORDS);
+
+  tile.bbox = TRANSFORM_TO_WORLD_BBOX;
+  tile.transformToWorld = content =>
+    content.map(object => _transformTileCoordsToWGS84(object, tile, TRANSFORM_TO_WORLD_VIEWPORT));
+  // Mock loader
+  tile._loader = getTileData();
+  // Not loaded data
+  t.is(
```
Comment (Pessimistress) [inline test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:129]: You need to test `await tile.dataInWorldCoordinates`

File: modules/geo-layers/src/index.js [hunk 25]
```diff
@@ -27,4 +27,7 @@ export {default as H3ClusterLayer} from './h3-layers/h3-cluster-layer';
 export {default as H3HexagonLayer} from './h3-layers/h3-hexagon-layer';
 export {default as Tile3DLayer} from './tile-3d-layer/tile-3d-layer';
 export {default as TerrainLayer} from './terrain-layer/terrain-layer';
-export {default as MVTLayer} from './mvt-layer/mvt-layer';
+export {
```
Comment (Pessimistress) [inline modules/geo-layers/src/index.js:30]: Don't export things just for testing purpose. You can import directly from `@deck.gl/geo-layers/mvt-layer/mvt-layer` in a spec file.

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js (hunk 24 above)
Comment (AdriSolid) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:129]: In case data is not loaded `dataInWorldCoordinates` returns a promise, can we check it like this, right?
```javascript
t.is(
    tile.dataInWorldCoordinates instanceof Promise,
    tile._loader instanceof Promise,
    'dataInWorldCoordinates getter correctly returns loader.'
  );
```

Getting data from getter:
```javascript
t.deepEqual(
    await tile.dataInWorldCoordinates,
    tile._transformToWorldDataCache,
    'dataInWorldCoordinates getter correctly returns cached data.'
  );
```

Comment (AdriSolid) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:129]: You are right, checking

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 26]
```diff
@@ -65,3 +67,78 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+      ]
+    }
+  }
+];
+
+test('Tile2DHeader#dataInWorldCoordinates', async t => {
+  const [loader, tile] = createTile2DHeaderWithMockedData(LOCAL_TILE_COORDS);
+
+  t.notOk(tile.isLoaded, 'Data has not been loaded yet');
+  t.deepEqual(await tile.dataInWorldCoordinates, await loader);
```
Comment (Pessimistress) [inline test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:92]: Delete this line please. You are still cheating around your bug. When you call `await tile.dataWorldCoordinates` twice they return different results.

File: modules/geo-layers/src/mvt-layer/mvt-layer.js [hunk 27]
```diff
@@ -217,12 +218,29 @@ export default class MVTLayer extends TileLayer {
...
   }
 
+  getVisibleTiles() {
+    if (this.state && this.state.tileset && this.state.tileset.isLoaded) {
+      return this.state.tileset.selectedTiles;
+    }
+    return [];
+  }
+
+  _setupTileTransform(tiles, viewport) {
```
Comment (alasarr) [inline modules/geo-layers/src/mvt-layer/mvt-layer.js:228]: Use a pure function instead of a class method

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js (hunk 26 above)
Comment (AdriSolid) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:92]: True, didn't see it..

Comment (Pessimistress) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js]: The ultimate problem in these tests are that, you are testing around your own code (is my code executing the way I write it?) instead of testing a real world scenario (is the user getting what they expect when they use this API?). While both will give you the same coverage numbers, the latter is always more meaningful.

Comment (Pessimistress) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js]: For example, your user will never call `await data._loader` themselves, so make a test assertion with that value is useless. In this case. The await also has a side effect (changes what `tile.dataInWorldCoordinates` returns), so by doing it you also invalidated any following up test.

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 28]
```diff
@@ -65,3 +67,78 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+  const requestScheduler = new RequestScheduler({throttleRequests: true, maxRequests: 1});
+
+  const getTileData = () => Promise.resolve(data);
+  const onTileLoad = () => null;
+  const onTileError = () => null;
+
+  const tile = new Tile2DHeader({onTileLoad, onTileError});
+  tile.isSelected = true;
+
+  const loader = tile._loadData(getTileData, requestScheduler);
```
[the remark this answers is not in the document]
Comment (AdriSolid) [reply test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:145]: I would keep calling `tile._loadData`, even if it's a private method, since the [other implemented tests](https://github.com/visgl/deck.gl/blob/013380268dcf3fb4a2b0d4000f2ac2e464c8935f/test/modules/geo-layers/tile-layer/tile-2d-header.spec.js#L41) are using this method, so would need to modify them..

**Review by Pessimistress — CHANGES_REQUESTED** [review]
Blocking until `tile.dataInWorldCoordinates` is fixed

File: modules/geo-layers/src/tile-layer/tile-2d-header.js [hunk 29]
```diff
@@ -17,12 +18,29 @@ export default class Tile2DHeader {
...
+    this.transformToWorld = null;
+    this._transformToWorldDataCache = null;
   }
 
   get data() {
     return this._isLoaded ? this.content : this._loader;
   }
 
+  get dataInWorldCoordinates() {
+    if (this.transformToWorld) {
```
Comment (alasarr) [inline modules/geo-layers/src/tile-layer/tile-2d-header.js:31]: For me, this is a mistake, if it's called and transformToWorld has not been provided we should just to throw an exception

File: docs/api-reference/geo-layers/mvt-layer.md [hunk 30]
```diff
@@ -161,6 +158,41 @@ new MVTLayer({
 })
 ```
 
+* `getVisibleTiles` (Function)
```
Comment (alasarr) [inline docs/api-reference/geo-layers/mvt-layer.md:161]: We need to explain better this method, we need to explain what the tile object has inside. Something like [this](https://deck.gl/docs/api-reference/geo-layers/tile-layer#gettiledata)

File: test/modules/geo-layers/mvt-layer.spec.js [hunk 31]
```diff
@@ -243,9 +245,42 @@ test('TileJSON', async t => {
...
   await testLayerAsync({Layer: TestMVTLayer, viewport: testViewport, testCases, onError: t.notOk});
+
   t.end();
 
-  // restore fetcch
+  // restore fetch
   _global.fetch = fetch;
 });
+
+test('MVT#getVisibleTiles', async t => {
```
Comment (alasarr) [inline test/modules/geo-layers/mvt-layer.spec.js:257]: I'd change this test to something more real. For example, set the viewport to the whole world and check you have the 4 expected tiles at z=1

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 32]
```diff
@@ -65,3 +67,82 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+  const [loader, tile] = createTile2DHeaderWithMockedData(LOCAL_TILE_COORDS);
+
+  t.notOk(tile.isLoaded, 'Data has not been loaded yet');
+
+  await loader;
+
+  t.ok(tile.isLoaded, 'Data has been loaded');
+  t.deepEqual(
+    await tile.dataInWorldCoordinates,
+    tile.content,
```
Comment (alasarr) [inline test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:98]: Check that it throws an exception if no transformWorld is provided.

File: test/modules/geo-layers/tile-layer/tile-2d-header.spec.js [hunk 33]
```diff
@@ -65,3 +67,82 @@ test('Tile2DHeader#Abort quickly', async t => {
...
+
+  t.notOk(tile.isLoaded, 'Data has not been loaded yet');
+
+  await loader;
+
+  t.is(tile._transformToWorldDataCache, null, 'Data cache is empty');
+  t.ok(tile.isLoaded, 'Data has been loaded');
+  t.deepEqual(
+    await tile.dataInWorldCoordinates,
+    tile._transformToWorldDataCache,
```
Comment (alasarr) [inline test/modules/geo-layers/tile-layer/tile-2d-header.spec.js:128]: We're just checking cache is working but not if the response is correct.

Instead of doing that, let's create an object with the expected return and compare that...And after that, check that cache is set properfy, and if you call again check transformToWorld is not called again (because of the cache) and the results are still the same

**Review by alasarr — CHANGES_REQUESTED** [review]

**Comment by alasarr** [comment]
After talking with @AdriSolid and thanks to the comments/help of @Pessimistress, we've decided to change the approach.

I didn't see too much sense in the original API, because we provided the getVisibleTiles method on the viewPortChanges callback. However, we already have a callback for TileLayer which does exactly does:

>onViewportLoad is a function that is called when all tiles in the current viewport are loaded. The loaded content (as returned by getTileData) for each visible tile is passed as an array to this callback function.

However, the callback at TileLayer doesn't return the x,y,z of the tile and it's quite difficult to work with this data. For example, if the user wants to transform to WGS84 we need to provide the reference tile.

As onViewportLoad guarantees, all the tiles have been loaded the code is simpler for the user.

```javascript
const onViewportLoad = tiles => {
  tiles.forEach(tile => {
    // data in local coordinates
    const dataInLocalCoords = tile.data;
    // data in WGS84
    const dataInWGS84 = tile.dataInWorldCoordinates;
  });
};

new MVTLayer({
  id: "..."
  data: "..."
  onViewportLoad
})
```

**Comment by alasarr** [comment]
Closing this in favor of https://github.com/visgl/deck.gl/pull/5360
